public/app/plugins/panel/barchart/BarChartTooltip.tsx

```diff
@@ -0,0 +1,177 @@
+import { css } from '@emotion/css';
+import React from 'react';
+
+import {
+  DataFrame,
+  FALLBACK_COLOR,
+  FieldType,
+  GrafanaTheme2,
+  getDisplayProcessor,
+  LinkModel,
+  Field,
+  getFieldDisplayName,
+  arrayUtils,
+} from '@grafana/data';
+import { SortOrder, TooltipDisplayMode } from '@grafana/schema/dist/esm/common/common.gen';
+import { useStyles2, useTheme2 } from '@grafana/ui';
+import { VizTooltipContent } from '@grafana/ui/src/components/VizTooltip/VizTooltipContent';
+import { VizTooltipFooter } from '@grafana/ui/src/components/VizTooltip/VizTooltipFooter';
+import { VizTooltipHeader } from '@grafana/ui/src/components/VizTooltip/VizTooltipHeader';
+import { ColorIndicator, ColorPlacement, LabelValue } from '@grafana/ui/src/components/VizTooltip/types';
+
+import { getDataLinks } from '../status-history/utils';
+
+// exemplar / annotation / time region hovering?
+// add annotation UI / alert dismiss UI?
+
+interface TimeSeriesTooltipProps {
+  frames?: DataFrame[];
+  // aligned series frame
+  seriesFrame: DataFrame;
+  // hovered points
+  dataIdxs: Array<number | null>;
+  // closest/hovered series
+  seriesIdx?: number | null;
+  mode?: TooltipDisplayMode;
+  sortOrder?: SortOrder;
+
+  isPinned: boolean;
+
+  annotate?: () => void;
+}
+
+export const BarChartTooltip = ({
+  frames,
+  seriesFrame,
+  dataIdxs,
+  seriesIdx,
+  mode = TooltipDisplayMode.Single,
+  sortOrder = SortOrder.None,
+  isPinned,
+  annotate,
+}: TimeSeriesTooltipProps) => {
+  const theme = useTheme2();
+  const styles = useStyles2(getStyles);
+
+  const xField = seriesFrame.fields[0];
+  if (!xField) {
+    return null;
+  }
+
+  const xFieldFmt = xField.display || getDisplayProcessor({ field: xField, theme });
+  let xVal = xFieldFmt(xField!.values[dataIdxs[0]!]).text;
+
+  let links: Array<LinkModel<Field>> = [];
+  let contentLabelValue: LabelValue[] = [];
+
+  // Single mode
+  if (mode === TooltipDisplayMode.Single) {
```

Comment:
Would be nice to see sections commented as this is fairly dense code, it's good stuff though!